kale/interpret/correlation_analysis.py

```diff
@@ -0,0 +1,144 @@
+"""
+Correlation of uncertainty with error (fit_line_with_ci)
+"""
+from typing import Any, Dict, List, Optional
+
+import matplotlib.pyplot as plt
+import numpy as np
+import pwlf
+from matplotlib.ticker import ScalarFormatter
+from scipy import stats
+
+
+def fit_line_with_ci(
+    errors: np.ndarray,
+    uncertainties: np.ndarray,
+    quantile_thresholds: List[float],
+    cmaps: List[Dict[str, str]],
+    to_log: bool = False,
+    error_scaling_factor: float = 1.0,
+    save_path: Optional[str] = None,
+) -> Dict[str, List[Any]]:
+    """
+    Calculates Spearman correlation between errors and uncertainties.
+    Plots piecewise linear regression with bootstrap confidence intervals.
+    Breakpoints in linear regression are defined by the uncertainty quantiles of the data.
+
+    Args:
+        errors (np.ndarray): Array of errors.
+        uncertainties (np.ndarray): Array of uncertainties.
+        quantile_thresholds (List[float]): List of quantile thresholds.
+        cmaps (List[Dict[str, str]]): List of colormap dictionaries.
+        to_log (bool, optional): Whether to apply logarithmic transformation on axes. Defaults to False.
+        error_scaling_factor (float, optional): Scaling factor for error. Defaults to 1.0.
+        save_path (Optional[str], optional): Path to save the plot, if None, the plot will be shown. Defaults to None.
+
+    Returns:
+        Dict[str, List[Any]]: Dictionary containing Spearman and Pearson correlation coefficients and p-values.
+    """
+
+    plt.figure(figsize=(12, 8))
+    ax = plt.gca()
+
+    if to_log:
+        plt.xscale("log", base=2)
+        plt.yscale("log", base=2)
+
+    X = uncertainties
+    y = errors * error_scaling_factor
+    plt.xlim(max(X), min(X))
+
+    # Calculate correlations
+    spm_corr, spm_p = stats.spearmanr(X, y, alternative="greater")
+    pear_corr, pear__p = stats.pearsonr(X, y)
+    correlation_dict = {"spearman": [spm_corr, spm_p], "pearson": [pear_corr, pear__p]}
+
+    # Bootstrap sample for confidence intervals
+    for i in range(0, 1000):
+        sample_index = np.random.choice(range(0, len(y)), int(len(y) * 0.6))
+
+        X_samples = X[sample_index]
+        y_samples = y[sample_index]
+        my_pwlf = pwlf.PiecewiseLinFit(X_samples, y_samples)
+        my_pwlf.fit_with_breaks(quantile_thresholds)
+        xHat = np.linspace(min(X), max(X), num=10000)
+        yHat = my_pwlf.predict(xHat)
+
+        # Plot each bootstrap as a grey line
+        plt.plot(xHat[::-1], yHat[::-1], "-", color="grey", alpha=0.2, zorder=2)
+
+    # Display all datapoints on plot
+    plt.scatter(X, y, marker="o", color=cmaps[2], zorder=1, alpha=0.2)
```

Comment:
Suggest making the colour more explicit instead of `cmaps[2]`. It's not obvious what the size or length of `cmaps` is.